File: baseunits/modules/MangaInn.pas
unit MangaInn;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, WebsiteModules, uData, uBaseUnit, uDownloadsManager,
  XQueryEngineHTML, synautil, FMDVars;

implementation

function GetNameAndLink(const MangaInfo: TMangaInformation;
  const ANames, ALinks: TStringList; const AURL: String;
  const Module: TModuleContainer): Integer;
var
  s: String;
begin
  Result := NET_PROBLEM;
  s := ALPHA_LIST[Module.CurrentDirectoryIndex + 1];
  if MangaInfo.FHTTP.GET(Module.RootURL + '/manga-list/' + s) then
  begin
    Result := NO_ERROR;
    with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do
      try
        updateList.CurrentDirectoryPageNumber := 1;
        XPathHREFAll('//ul[contains(@class, "manga-list")]/li/a', ALinks, ANames);
      finally
        Free;
      end;
  end;
end;

function GetInfo(const MangaInfo: TMangaInformation;
  const AURL: String; const Module: TModuleContainer): Integer;
var
  v: IXQValue;
begin
  Result := NET_PROBLEM;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  with MangaInfo.mangaInfo, MangaInfo.FHTTP do
  begin
    url := MaybeFillHost(Module.RootURL, AURL);
    if GET(url) then begin
      Result := NO_ERROR;
      with TXQueryEngineHTML.Create(Document) do
        try
          coverLink := XPathString('//div[contains(@class, "manga-info")]//img[@class="img-responsive mobile-img"]/resolve-uri(@src)');
          title := XPathString('//title');
          if Pos(' - Read ', title) <> 0 then
            title := Trim(GetBetween('- Read ', ' Online at', title));
          status := MangaInfoStatusIfPos(
            XPathString('//dt[starts-with(.,"Status")]/following-sibling::dd[1]'),
            'Ongoing',
            'Complete');
          authors := XPathString('//dt[starts-with(.,"Author")]/following-sibling::dd[1]');
          artists := XPathString('//dt[starts-with(.,"Artist")]/following-sibling::dd[1]');
          genres := XPathStringAll('//dt[starts-with(.,"Categories")]/following-sibling::dd[1]/a');
          summary := Trim(XPathString('//div[contains(@class, "manga-info")]//div[contains(@class, "note")]'));
          for v in XPath('//div[@id="chapter_list"]/ul/li/a') do
          begin
            chapterLinks.Add(v.toNode().getAttribute('href'));
            chapterName.Add(XPathString('span[1]', v));
          end;
          InvertStrings([chapterLinks, chapterName]);
        finally
          Free;
        end;
    end;
  end;
end;

function GetPageNumber(const DownloadThread: TDownloadThread;
  const AURL: String; const Module: TModuleContainer): Boolean;
var
  s: String;
begin
  Result := False;
  if DownloadThread = nil then Exit;
  with DownloadThread.Task.Container, DownloadThread.FHTTP do
  begin
    PageLinks.Clear;
    PageNumber := 0;
    if GET(MaybeFillHost(Module.RootURL, AURL + '/all-pages')) then
    begin
      Result := True;
      with TXQueryEngineHTML.Create(Document) do
        try
          s := XPathString('//script[contains(.,"var images = [")]');
          if s <> '' then
          begin
            s := '[' + GetBetween('var images = [', '];', s) + ']';
            ParseHTML(s);
            XPathStringAll('json(*)().url', PageLinks);
          end;
        finally
          Free;
        end;
    end;
  end;
end;

procedure RegisterModule;
begin
  with AddModule do
  begin
    Website := 'MangaInn';
    RootURL := 'http://www.mangainn.net';
    Category := 'English';
    OnGetNameAndLink := @GetNameAndLink;
    OnGetInfo := @GetInfo;
    OnGetPageNumber := @GetPageNumber;
    TotalDirectory := Length(ALPHA_LIST);
  end;
end;

initialization
  RegisterModule;

end.

File: lua/modules/MangaInn.lua
----------------------------------------------------------------------------------------------------
-- Scripting Parameters
----------------------------------------------------------------------------------------------------

local LuaDebug   = require 'LuaDebugging'
-- LuaDebugging  = true   --> Override the global LuaDebugging variable by uncommenting this line.
-- LuaStatistics = true   --> Override the global LuaStatistics variable by uncommenting this line.


----------------------------------------------------------------------------------------------------
-- Local Constants
----------------------------------------------------------------------------------------------------

local AlphaList = '#abcdefghijklmnopqrstuvwxyz'
local DirectoryPagination = '/manga-list/'
local LoadAllImages = '/all-pages'


----------------------------------------------------------------------------------------------------
-- Event Functions
----------------------------------------------------------------------------------------------------

-- Get info and chapter list for current manga.
function GetInfo()
  local v, x = nil
  local u = MaybeFillHost(module.RootURL, url)
  
  --[[Debug]] LuaDebug.WriteLogWithHeader('GetInfo', 'url ->  ' .. u)
  if not http.Get(u) then return net_problem end
  
  x = TXQuery.Create(http.Document)
  mangainfo.Title     = x.XPathString('(//h5[@class="widget-heading"])[1]')
  mangainfo.CoverLink = x.XPathString('//div[@class="content"]//img/@src')
  mangainfo.Status    = x.XPathString('//dt[contains(., "Status")]/following-sibling::dd[1]')
  mangainfo.Authors   = x.XPathString('//dt[contains(., "Author")]/following-sibling::dd[1]')
  mangainfo.Artists   = x.XPathString('//dt[contains(., "Artist")]/following-sibling::dd[1]')
  mangainfo.Genres    = x.XPathStringAll('//dt[contains(., "Categories")]/following-sibling::dd[1]/a')
  mangainfo.Summary   = Trim(x.XPathString('//div[contains(@class, "note")]'))
  
  v = x.XPath('//div[@id="chapter_list"]/ul/li/a')
  for i = 1, v.Count do
    mangainfo.ChapterLinks.Add(v.Get(i).GetAttribute('href'))
    mangainfo.ChapterNames.Add(x.XPathString('span[1]', v.Get(i)))
  end
  InvertStrings(mangainfo.ChapterLinks, mangainfo.ChapterNames)
  
  --[[Debug]] LuaDebug.PrintMangaInfo()
  --[[Debug]] LuaDebug.WriteStatistics('Chapters', mangainfo.ChapterLinks.Count .. '  (' .. mangainfo.Title .. ')')
  
  return no_error
end


-- Get links and names from the manga list of the current website.
function GetNameAndLink()
  local x = nil
  local u = module.RootURL .. DirectoryPagination .. AlphaList:sub(module.CurrentDirectoryIndex + 1, module.CurrentDirectoryIndex + 1)
  
  --[[Debug]] LuaDebug.WriteLogWithHeader('GetNameAndLink', 'url ->  ' .. u)
  if not http.Get(u) then return net_problem end
  
  x = TXQuery.Create(http.Document)
  x.XPathHREFAll('//ul[contains(@class, "manga-list")]/li/a', links, names)
  updatelist.CurrentDirectoryPageNumber = 1
  
  --[[Debug]] LuaDebug.PrintMangaDirectoryEntries(module.CurrentDirectoryIndex + 1)
  
  return no_error
end


-- Get the page count for the current chapter.
function GetPageNumber()
  local s, x = nil
  local u = MaybeFillHost(module.RootURL, url) .. LoadAllImages
  
  --[[Debug]] LuaDebug.WriteLogWithHeader('GetPageNumber', 'url ->  ' .. u)
  if not http.Get(u) then return net_problem end
  
  x = TXQuery.Create(http.Document)
  x.ParseHTML(GetBetween('var images = ', ';', x.XPathString('//script[contains(., "var images = ")]')):gsub('\\/', '/'))
  x.XPathStringAll('json(*)().url', task.PageLinks)
  
  --[[Debug]] LuaDebug.PrintChapterPageLinks()
  --[[Debug]] LuaDebug.WriteStatistics('ChapterPages', task.PageLinks.Count .. '  (' .. u .. ')')
  
  return no_error
end


-- Prepare the url, http header and/or http cookies before downloading an image.
function BeforeDownloadImage()
  http.Headers.Values['referer'] = module.RootURL
  
  return true
end


----------------------------------------------------------------------------------------------------
-- Module Initialization
----------------------------------------------------------------------------------------------------

function Init()
  AddWebsiteModule('MangaInn', 'http://www.mangainn.net', 'English')
end

function AddWebsiteModule(name, url, category)
  local m = NewModule()
  m.Website               = name
  m.RootURL               = url
  m.Category              = category
  m.TotalDirectory        = AlphaList:len()
  m.OnGetInfo             = 'GetInfo'
  m.OnGetNameAndLink      = 'GetNameAndLink'
  m.OnGetPageNumber       = 'GetPageNumber'
  m.OnBeforeDownloadImage = 'BeforeDownloadImage'
  return m
end